Source/WebCore/page/PointerLockController.cpp

```diff
@@ -40,44 +42,83 @@
+    }
+};
+
 PointerLockController::PointerLockController(Page& page)
     : m_page(page)
 {
 }
 
-void PointerLockController::requestPointerLock(Element* target)
+PointerLockController::~PointerLockController()
+{
+}
+
+void PointerLockController::requestPointerLock(Element* target, bool unadjustedMovement, RefPtr<DeferredPromise> promise)
 {
     if (!target || !target->isConnected() || m_documentOfRemovedElementWhileWaitingForUnlock) {
         enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        if (promise)
+            promise->reject(WrongDocumentError);
         return;
     }
 
     if (m_documentAllowedToRelockWithoutUserGesture != &target->document() && !UserGestureIndicator::processingUserGesture()) {
         enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        // If the request was not started from an engagement gesture and the Document has not previously released a successful Pointer Lock with exitPointerLock():
+        if (promise)
+            promise->reject(NotAllowedError);
         return;
     }
 
     if (target->document().isSandboxed(SandboxPointerLock)) {
         // FIXME: This message should be moved off the console once a solution to https://bugs.webkit.org/show_bug.cgi?id=103274 exists.
         target->document().addConsoleMessage(MessageSource::Security, MessageLevel::Error, "Blocked pointer lock on an element because the element's frame is sandboxed and the 'allow-pointer-lock' permission is not set."_s);
         enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        // If this's node document's active sandboxing flag set has the sandboxed pointer lock browsing context flag set:
+        if (promise)
+            promise->reject(SecurityError);
+        return;
+    }
+
+    if (unadjustedMovement && !m_page.chrome().client().pointerLockSupportsUnadjustedMovement()) {
+        // If options["unadjustedMovement"] is true and the platform does not support unadjustedMovement:
+        enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        if (promise)
+            promise->reject(NotSupportedError);
         return;
     }
 
     if (m_element) {
         if (&m_element->document() != &target->document()) {
             enqueueEvent(eventNames().pointerlockerrorEvent, target);
+            // If the user agent's pointer-lock target is an element whose shadow-including root is not equal to this's shadow-including root, then:
+            if (promise)
+                promise->reject(InvalidStateError);
             return;
         }
         m_element = target;
+        m_unadjustedMovement = unadjustedMovement;
         enqueueEvent(eventNames().pointerlockchangeEvent, target);
+        if (promise)
+            promise->resolve();
         m_page.pointerCaptureController().pointerLockWasApplied();
     } else {
         m_lockPending = true;
         m_element = target;
+        m_unadjustedMovement = unadjustedMovement;
+        m_promise = promise;
```

Comment:
Why do we only update `m_promise` in this path?

Reply:
All other paths either resolve or reject the promise. This is the only path where we have to wait until called back by either didAcquirePointerLock/didNotAcquirePointerLock to resolve or reject the promise later.

Reply:
If `requestPointerLock` is called back to back then the second promise gets resolved immediately and not after pointer lock was actually started. It's not entirely a new issue because this issue was already present with the event path as well and `pointerlockchange` gets sent early there.

And then, if `m_lockPending` is true, but then `didNotAcquirePointerLock` happens, then the second promise is resolved, but the first one is rejected. Both should be rejected. In practice, the easiest way to achieve this is to modify MiniBrowser to resolve `_webViewDidRequestPointerLock:completionHandler:`, with NO, but I think it could happen in Safari if window focus is lost at the right time or if pointer lock had been enabled before.

So yeah, it's broken. FWIW, it's even more broken in Chromium, where just calling `requestPointerLock` twice in a row without user activation in Chromium is enough to get it to incorrectly send `pointerlockchange` and to resolve the second promise (WebKit passes here because it checks User Activation synchronously).

I think the right thing to do when `PointerLockController::lockPending` is to keep a vector of pending promises and either resolve or reject them in order that they were created depending on what eventually happens with `did(Not)AcquirePointerLock`, and similarly only send a `pointerlockchange` or `pointerlockerror` event once we know what happens.

The other approach I can see would be to reject `m_promise` if it exists when re-locking because `lockPending` is true, but this seems wrong because just because a pointer lock is immediately replaced by another in the same document doesn't mean that the first one failed, it's just that it was active for zero time.

I'll update the patch accordingly.

Exhibit A - Broken when `_webViewDidRequestPointerLock:completionHandler:` calls `handler(NO)`:

```html
<!DOCTYPE html>
<html>
<body>
<div id="target1"></div>
<button id="lock">Lock</button>
</body>
<script>
let el = document.getElementById('target1');
let btn = document.getElementById('lock');
document.addEventListener('pointerlockchange', () => { console.log('pointerlockchange'); });
document.addEventListener('pointerlockerror', () => { console.log('pointerlockerror'); });
btn.addEventListener('click', () => {
    let p1 = el.requestPointerLock();
    let p2 = el.requestPointerLock();
    p1.then(() => console.log("p1 resolved")).catch(ex => console.error('p1 caught', ex));
    p2.then(() => console.log("p2 resolved")).catch(ex => console.error('p2 caught', ex));
});
</script>
</html>
```

Exhibit B: Broken in Chromium, but works in WebKit:

```html
<!DOCTYPE html>
<html>
<body>
<div id="target1"></div>
</body>
<script>
let el = document.getElementById('target1');
document.addEventListener('pointerlockchange', () => { console.log('pointerlockchange'); });
document.addEventListener('pointerlockerror', () => { console.log('pointerlockerror'); });
let p1 = el.requestPointerLock(); // should fail because no user activation.
let p2 = el.requestPointerLock(); // unexpectedly succeeds on Chromium
p1.then(() => console.log("p1 resolved")).catch(ex => console.error('p1 caught', ex));
p2.then(() => console.log("p2 resolved")).catch(ex => console.error('p2 caught', ex));
</script>
</html>
```

Reply:
Thanks for the elaborate explanation here! I think I had only one question here -- in which direction do you want to update your patch? Do you want to maintain ordering of the pending promises?

Reply:
I think it would make most sense to resolve the promises in the order corresponding to the requestPointerLock calls that created them. So in Exhibit A above, it should print:

```
p1 resolved
p2 resolved
```

The reason for this is suppose p2 is for a requestPointerLock on a different element than p1 was (but still same document). If we resolved p2 first, it might lead to confusion about which element is going to be receiving the events (not internally in WebKit, but I mean web page authors might be confused -- I know I would be confused if it were the other way around).

Reply:
This is done in the latest patch and I added a test for it. The test doesn’t mandate any specific order for the promises to resolve/reject but it does at least require that they are resolved/rejected consistently.